### PR TITLE
Corrected indexing bug when fitting hyperparameters in NeuralNet.

### PR DESCRIPTION
**Pull request by zakv**
Changes proposed in this pull request:

- Fix indexing when dividing up data into a training set and cross validation set. This is done while fitting hyperparameters in `NeuralNet.fit_neural_net()`.
  - Previously 10% of the data would only be used for training, 10% would only be used for cross validation, and 80% would be used for both.

Things are set up in such a way that `self.fit_hyperparameters` is `False` for all instances of `NeuralNet` so the section of the code with this bug is never reached in normal operation. I was considering playing around with `self.fit_hyperparameters` though so I figured I'd submit this fix.

@qctrl/support @charmasaur

### CODE REVIEW

File: mloop/neuralnet.py [hunk 1]
```diff
@@ -648,8 +648,8 @@ def fit_neural_net(self, all_params, all_costs):
                 cv_size = int(len(all_params) / 10)
                 train_params = all_params[:-cv_size]
                 train_costs = all_costs[:-cv_size]
-                cv_params = all_params[cv_size:]
-                cv_costs = all_costs[cv_size:]
+                cv_params = all_params[-cv_size:]
```
Comment (charmasaur) [inline mloop/neuralnet.py:651]: Haha, damn it. Maybe this is why I never had much success with this method!

That said, what do you think about changing to
```
cv_params = all_params[:cv_size]
cv_costs = all_params[:cv_size]
train_params = all_params[cv_size:]
train_costs = all_params[cv_size:]
```
just to save messing around with negative indices?

Comment (zakv) [reply mloop/neuralnet.py:651]: Hmm, that's got me thinking actually. Do you think it would be better to use the most recent parameters in the cross validation set? Generally the later parameters are likely to be closer to the optimal values than the earlier ones. That's the region of parameter space where we'd care most about the effects of overfitting, so I think that's the region of parameter space where the cross validation data points would ideally be.

Put another way, I suspect that if the net is trained mostly on data points near the optimum, then it wouldn't be ideal to assess its "overfitted-ness" at some data points in a potentially far-off region of parameter space.

To avoid negative indices we could still do something like this:
```python
training_fraction = 0.9
split_index = int(training_fraction * len(all_params))
train_params = all_params[:split_index]
train_costs = all_params[:split_index]
cv_params = all_params[split_index:]
cv_costs = all_params[split_index:]
```

I'll implement something like that for now, but I'd be happy to hear your thoughts and wouldn't at all mind changing things.

**Comment by zakv** [comment]
I also just noticed that the logic for fitting hyperparameters every 20th fit seems a bit off as well. The expression `int(len(all_params + 5) / 20)` is missing parentheses and so it adds 5 to all of the values in `all_params` instead of adding 5 to its length. Also, to start after 5 iterations (as mentioned in the comment above that line) I think that should be `+ 15`. Lastly, I have a small nitpick about the comment. This actually keeps track of the number of runs, not the number of fits, which is different because of the training runs.

Since there isn't a particular reason to fit hyperparameters after 5 runs, I'm going to just set it to run every 20 runs for now so that things are a little simpler. Happy to change it back though if desired. See commit 21d971e.

File: mloop/neuralnet.py (hunk 1 above)
Comment (zakv) [reply mloop/neuralnet.py]: Implemented in 27157d5.

Comment (charmasaur) [reply mloop/neuralnet.py]: I don't have much intuition for this, although I agree that just taking cross validation parameters from the start (like I proposed) probably isn't ideal.

One concern with taking them from the end is that we might go too far the other way, and get a worse fit as a result of not training on any of the parameters closest to the optimum (minimising the cross validation loss will help, but it could be that the minimising fit is still kind of bad).

Still, what you've got definitely seems like a step in the right direction, so I'm happy to stick with that and iterate if necessary once we've seen how it works in action.

**Comment by charmasaur** [comment]
This is all very embarrassing, but thanks for the fixes!
